[CDAP-15787] add validation apis for gcs source and sink

CuriousVini opened the pull request:
Requires https://github.com/cdapio/hydrator-plugins/pull/947 changes to be merged.

albertshau commented on src/main/java/io/cdap/plugin/gcp/gcs/sink/GCSBatchSink.java:
```diff
@@ -106,19 +111,41 @@ protected void recordLineage(LineageRecorder lineageRecorder, List<String> outpu
+        try {
+          new SimpleDateFormat(suffix);
+        } catch (IllegalArgumentException e) {
+          collector.addFailure("Invalid suffix.", "Ensure provided suffix is valid.")
```
should we include the message from the exception? Otherwise it won't be apparent why it's invalid.

Also don't think we need the corrective action if we include the exception message

albertshau commented on src/main/java/io/cdap/plugin/gcp/gcs/sink/GCSMultiBatchSink.java:
```diff
@@ -62,24 +63,37 @@ public GCSMultiBatchSink(GCSMultiBatchSinkConfig config) {
-      throw new IllegalArgumentException(String.format("Could not find the '%s' output format plugin.",
-                                                       format.name().toLowerCase()));
+
+    if (format != null) {
```
don't think this check is needed. Instead, if the format couldn't be fetched above, we can just throw the exception there since everything below it requires that format is not null.

albertshau commented on src/main/java/io/cdap/plugin/gcp/gcs/sink/GCSBatchSink.java:
```diff
@@ -129,18 +128,10 @@ public void validate(FailureCollector collector) {
       }
-      if (!containsMacro(NAME_FORMAT)) {
-        try {
-          getFormat();
```
doesn't this still need to be called?

CuriousVini replied:
When would it be invalid? since its not macro enabled.

albertshau commented on src/main/java/io/cdap/plugin/gcp/gcs/source/GCSSource.java:
```diff
@@ -203,13 +201,6 @@ public void validate(FailureCollector collector) {
       }
-      if (!containsMacro(NAME_FORMAT)) {
-        try {
-          getFormat();
```
doesn't this still need to be called?

albertshau commented on src/main/java/io/cdap/plugin/gcp/gcs/source/GCSSource.java:
```diff
@@ -204,7 +229,7 @@ public Pattern getFilePattern() {
         return fileRegex == null ? null : Pattern.compile(fileRegex);
       } catch (RuntimeException e) {
-        throw new IllegalArgumentException("Invalid file regular expression: " + e.getMessage(), e);
+        throw new IllegalArgumentException("Invalid file regular expression.", e);
```
why remove the message? It will be hard to tell what to change without it.

albertshau commented on src/main/java/io/cdap/plugin/gcp/gcs/source/GCSSource.java:
```diff
@@ -169,18 +170,42 @@ public GCSSourceConfig() {
+        try {
+          getFilePattern();
+        } catch (IllegalArgumentException e) {
+          collector.addFailure("Provide a valid regular expression.", null)
```
this won't say what is wrong with the pattern, so the user won't be able to adjust accordingly

CuriousVini commented:
I removed exception messages because of this:
![image](https://user-images.githubusercontent.com/14131070/65657705-76783f00-dfd9-11e9-9f4d-e6b44ee495b0.png)

For file system properties, since it must be a valid json, not adding exception message for `getFileSystemProperties` call. Others I have added error message.

CuriousVini commented:
squashing and merging.
